README.md
# SimPermut
 
## Overview

This repository is for the addon SimPermut which make a string usable by SimulationCraft with all the different permutations for the items you have in your inventory. This makes comparing stuff easy.

Also handle talent permutation, relic comparison, crucible comparison. 

Latest updates also add talents permutations and Artifact generator, aswell as more control over the profiles created. 

Don't hesitate to go on the [SimcMinMax](https://discordapp.com/invite/tFR2uvK) Discord in the #simpermut-autosimc Channel to ask about specific stuff.

Get it on curse : https://mods.curse.com/addons/wow/254069-simpermut

## How Can I Use It?

>**/SimPermut** [X[+Y[+Z...]]]

- Select which items you want to sim.

- Add enchants and gems on items that don't have any (you can force replace enchant and gem by ticking the checkbox).

- Click Generate.

- Copy the text and paste in simc.

 

>**Note:**

>You can also add parameters to pre-select an entire slot.

>ie: /SimPermut finger+trinket

>Will auto-select every ring and trinket and do a permutation with default parameters. 


## AutoSimC Export

You can now export selected gear to an **AutoSimC** data (settings.ini) to externalize simc string generation. For a large number of items, it will be faster.

See [AutoSimC on GitHub](https://github.com/SimCMinMax/AutoSimC)



## Known issues and developement plan

- Fix Warforged TW items

-[Rework code]

## Credits
Kutikuti 

[Watcheur](https://github.com/watcheur) for BFA changes

changelog.txt
v1.15.0 16/08/2018
[UI]Remove tier list
[UI]Remove legendaries
[UI]Adding Main hand & off hand
[UI]Add new enchantments
[UI]Add new gems
[Fix]Set minimum ilvl to 300 (was 200) & maximum to 500 (was 1000)
[Code]Cleanup (legion features)

v1.14.15 02/01/2018
[Fix]LUA error in some cases

v1.14.14 01/17/2018
[Fix]Fix artifact string export issue caused by 7.3.5 changes

v1.14.13 01/04/2018
[UI]Show T20 and T21 on main screen. T19 still accessible in the option panel
[Fix]Fix Amanthul detection for legendaries 

v1.14.12 12/12/2017
[Fix]Fix NC issue with off hand

v1.14.11 11/29/2017
[UI]updtate ilvl max threshold to 1k
[UI]Set ilvl max to 1k if it was at 999

v1.14.10 11/08/2017
[UI]Auto-simc format change
[Fix]Exclude amanthul from leg count

v1.14.9 11/03/2017
[UI]Auto-simc format change
[Fix]Shadow priest relic fix

v1.14.8 09/20/2017
[UI]New crucible export choice : export all combination of the current shown tree
[UI]Keep window position when changing module
[Code]Implement new crucible export format
[Fix]Reset crucible string when switching between one tree / full weapon

v1.14.7 09/13/2017
[Fix]Fix Crucible export

v1.14.6 09/13/2017
[Fix]Fix Crucible export
[Code]Cleanup

v1.14.5 09/04/2017
[UI]Added an option for 1 main stat gem
[Fix]Fix NC T2 icon

v1.14.4 09/04/2017
[UI]added more crucible options
[UI]added 7.3 gems
[Fix]added t21 to autosimc export

v1.14.3 08/31/2017
[UI]Added an option for small UI (option panel)
[UI]Added a way to add simc parameters (option panel)
[Fix]Fix crucible stuff
[FIX]Added some debug text for T21
[Code]cleanup

v1.14.1 08/30/2017
[UI]Netherlight Crucible generator
[UI]Add T21 (only in the option frame)
[Fix]Fixed a number of wrongly labelled relics
[Code]toc update

v1.13.3 08/08/2017
[Fix]Fixed size of an otpion that would sometimes cut itself
[Fix]Fixed timewalking items

v1.13.2 07/06/2017
[Fix]Fix autosimc export

v1.13.1 07/06/2017
[UI]Added an option to add gear from a link to permut
[FIX]Fix copy names with weird caracters
[FIX]Fix items with multiple gem slots
[FIX]Fix ilvl threshold for equiped items
[Code]Added security to some places where you get lua error in some rare cases

v1.12.5 06/28/2017
[FIX]Corrected artifact in off-hand
[UI]Added TOS relics

v1.12.4 06/07/2017
[UI]Added T20 selection
[UI]Added tier print for autosimc export
[FIX]Option pannel error with replace checkbox

v1.12.3 05/31/2017
[UI]Added an artifact generator/comparison
[UI]Added more classes to relic comparison
[UI]Added report type selection near generate button
[UI]Added default report types for relics and talents in option pannel
[CODE]Update artifact lib
[CODE]Added some more security check for atifact
[CODE]Added some more security check when 3rd relic slot is not unlocked
[FIX]Show an error when no profiles printed

v1.11.1 05/13/2017
[FIX]Fix generation

v1.11 05/12/2017
[UI]Added command line support to pre-select an entire slot
[UI]Added an option panel to personalize default parameters
[FIX]Fixed Heirloom not listing

v1.10.1 05/06/2017
[FIX]Fixed talent permutation
[UI]don't print current talents with the talent permut

v1.10 05/05/2017
[UI]Minor UI adjustment
[UI]Added an option to make a talent permutation

v1.9.2 04/22/2017
[UI]Minor UI adjustment
[UI]Auto set min leg with equiped legendaries
[FIX]Fix lua error in some rare cases


v1.9.1 04/19/2017
[UI]Added 0s in front of copie names so they are in the right order in simc

v1.9 04/18/2017
[UI]Added some UI option for leg min/max and T19

v1.8.3 04/07/2017
[UI]Small update to generate equiped gear directly after launch

v1.8.2 04/03/2017
[FIX]Fixed a bug with ring/trinket profile duplication detection

v1.8.1 03/29/2017
[FIX]Fix for 7.2 when you have unlocked new traits

v1.8 03/28/2017
[CODE]7.2 bump
[FIX]fix 7.2 api change
[Lib]updated LibArtifactData
[FIX]Fix artifact export for 7.2

v1.7.7-beta 03/28/2017
[FIX]Attempt to fix ring/trinket permutation

v1.7.6-beta 03/24/2017
[FIX]Fix lua error in debug mode

v1.7.5-beta 03/23/2017
[FIX]Follow yesterday's fix for ring/trinkets
[CODE]Added debugging stuff

v1.7.4 03/22/2017
[FIX]Fix useless lines printing even if not needed that was remvoed in last patch

v1.7.3 03/20/2017
[UI]Added focus to resultbox after generating

v1.7.2 03/17/2017
[FIX]Fixed inversion checker if only 2 rings/trinkets selected

v1.7.1 02/22/2017
[FIX]Corrected a small typo that broke inversion checker...
[CODE]Added some debug print

v1.7 02/21/2017
[UI]Put back inversion checker

v1.6.4 02/16/2017
[FIX]Fixed ring and trinket print if not selected

v1.6.3 02/15/2017
[FIX]Reduced Window size overall

v1.6.2 01/30/2017
[FIX]Fixed AutoSimc Export's artifact writing
[UI]Add an information when no copy are generated because there is no other valid combination
[UI]Integrated result window into main window for easier use

v1.6.1 01/23/2017
[FIX]Removed ring/trinket inversion check as it was causing too much problem that it solved. I'm working on a better solution
[FIX]Fixed a bug where ring/trinket would be added in generated chain even if they didn't change

v1.6-beta 01/19/2017
[UI]Added an AutoSimcExport button that generates a string of all item selected to export and do a permutation with AutoSimc
[UI]Added a var to select copy name type (number or item changed) in core.lua (ln 20). I will add a config panel later.
	report_type		= 1  -> list of item changed
	report_type		= 2  -> number (copyX)

v1.5.1 01/17/2017
[FIX]Fixed a bug when no ring/trinket where selected no generating the copy properly

v1.5 01/17/2017
[FIX]Removed a significant amount of copy due to ring/trinket inversion

v1.4 01/10/2017
[FIX]Removed the pieces that are from the base string from copy chain
[FIX]Removed impossible copy (2 leg max)

v1.3.1 12/16/2016
[FIX]Fixed copy name being written incorrectly
[FIX]Fixed Artifact update event incorrectly managed

v1.3 12/14/2016

libs/PersoLib.lua
local MAJOR, MINOR = "PersoLib", 1
local PersoLib = LibStub:NewLibrary(MAJOR, MINOR)

if not PersoLib then return end

local ArtifactUI          	= _G.C_ArtifactUI
local ArtifactRelicForgeUI  = _G.C_ArtifactRelicForgeUI
local ArtifactRelicForgeFrame = _G.ArtifactRelicForgeFrame
local HasArtifactEquipped 	= _G.HasArtifactEquipped
local SocketInventoryItem 	= _G.SocketInventoryItem
local ExtraData			

function PersoLib:Init(Datas)
	ExtraData=Datas
end

-- First letter in caps
function PersoLib:firstToUpper(str)
    return (str:gsub("^%l", string.upper))
end

-- permut item table
function PersoLib:doCartesianALACON(tableToPermut)
	local i1,i2,i3,i4,i5,i6,i7,i8,i9,i10,i11,i12,i13,i14
	local tableReturn={}
	
	for i1=1,#tableToPermut[1] do
		for i2=1,#tableToPermut[2] do
			for i3=1,#tableToPermut[3] do
				for i4=1,#tableToPermut[4] do
					for i5=1,#tableToPermut[5] do
						for i6=1,#tableToPermut[6] do
							for i7=1,#tableToPermut[7] do
								for i8=1,#tableToPermut[8] do
									for i9=1,#tableToPermut[9] do
										for i10=1,#tableToPermut[10] do
											for i11=1,#tableToPermut[11] do
												for i12=1,#tableToPermut[12] do
													for i13=1,#tableToPermut[13] do
														for i14=1,#tableToPermut[14] do
															tableReturn[#tableReturn+1]={}
															tableReturn[#tableReturn][1]=tableToPermut[1][i1]
															tableReturn[#tableReturn][2]=tableToPermut[2][i2]
															tableReturn[#tableReturn][3]=tableToPermut[3][i3]
															tableReturn[#tableReturn][4]=tableToPermut[4][i4]
															tableReturn[#tableReturn][5]=tableToPermut[5][i5]
															tableReturn[#tableReturn][6]=tableToPermut[6][i6]
															tableReturn[#tableReturn][7]=tableToPermut[7][i7]
															tableReturn[#tableReturn][8]=tableToPermut[8][i8]
															tableReturn[#tableReturn][9]=tableToPermut[9][i9]
															tableReturn[#tableReturn][10]=tableToPermut[10][i10]
															
															--ring
															tableReturn[#tableReturn][11]=tableToPermut[11][i11]
															tableReturn[#tableReturn][12]=tableToPermut[12][i12]
															
															--trinket
															tableReturn[#tableReturn][13]=tableToPermut[13][i13]
															tableReturn[#tableReturn][14]=tableToPermut[14][i14]
														end
													end
												end
											end
										end
									end
								end
							end
						end
					end
				end
			end
		end
	end
	
	
	
	return tableReturn
end

-- SimC, tokenize function
function PersoLib:tokenize(str)
  str = str or ""
  -- convert to lowercase and remove spaces
  str = string.lower(str)
  str = string.gsub(str, ' ', '_')
  str = string.gsub(str, ',', '_')

  -- strip trailing spaces
  if string.sub(str, str:len())=='_' then
    str = string.sub(str, 0, str:len()-1)
  end
  return str
end

function PersoLib:Split(str, delim, maxNb)
   -- Eliminate bad cases...
   if string.find(str, delim) == nil then
      return { str }
   end
   if maxNb == nil or maxNb < 1 then
      maxNb = 0    -- No limit
   end
   local result = {}
   local pat = "(.-)" .. delim .. "()"
   local nb = 0
   local lastPos
   for part, pos in string.gmatch(str, pat) do
      nb = nb + 1
      result[nb] = part
      lastPos = pos
      if nb == maxNb then
         break
      end
   end
   -- Handle the last field
   if nb ~= maxNb then
      result[nb + 1] = string.sub(str, lastPos)
   end
   return result
end

-- simc, method for constructing the talent string
function PersoLib:CreateSimcTalentString()
  local talentInfo = {}
  local maxTiers = 7
  local maxColumns = 3
  for tier = 1, maxTiers do
    for column = 1, maxColumns do
      local talentID, name, iconTexture, selected, available = GetTalentInfo(tier, column, GetActiveSpecGroup())
      if selected then
        talentInfo[tier] = column
      end
    end
  end

  local str = ''
  for i = 1, maxTiers do
    if talentInfo[i] then
      str = str .. talentInfo[i]
    else
      str = str .. '0'
    end
  end

  return str
end

-- simc, function that translates between the game's role values and ours
function PersoLib:translateRole(spec_id)
  local spec_role = ExtraData.RoleTable[spec_id]
  if spec_role ~= nil then
    return spec_role
  end
end

-- simc, function that reforulate race for simc
function PersoLib:getRace()
	-- Race info
	local _, playerRace = UnitRace('player')
	-- fix some races to match SimC format
	if playerRace == 'BloodElf' then
		playerRace = 'Blood Elf'
	elseif playerRace == 'NightElf' then
		playerRace = 'Night Elf'
	elseif playerRace == 'Scourge' then
		playerRace = 'Undead'
	end
	
	return playerRace
end

-- simc, function that returns spec ID
function PersoLib:getSpecID()
	local globalSpecID
	local specId = GetSpecialization()
	if specId then
		globalSpecID = GetSpecializationInfo(specId)
	end
	return globalSpecID
end

function PersoLib:IsArtifactFrameOpen()
	local ArtifactFrame = _G.ArtifactFrame
	return ArtifactFrame and ArtifactFrame:IsShown() or false
end

function PersoLib:IsCrucibleFrameOpen()
	local ArtifactRelicForgeFrame = _G.ArtifactRelicForgeFrame
	return ArtifactRelicForgeFrame and ArtifactRelicForgeFrame:IsShown() or false
end

function PersoLib:GetPowerData(powerId)
	if not powerId then
		return 0, 0
	end

	local powerInfo = ArtifactUI.GetPowerInfo(powerId)
	if powerInfo == nil then
		return powerId, 0
	end

	return powerId, powerInfo.currentRank - powerInfo.bonusRanks
end

function PersoLib:OpenArtifact()
	if not HasArtifactEquipped() then
		return false, false, 0
	end

	local artifactFrameOpen = PersoLib:IsArtifactFrameOpen()
	if not artifactFrameOpen then
		SocketInventoryItem(INVSLOT_MAINHAND)
	end

	local ArtifactFrame = _G.ArtifactFrame

	local itemId = select(1, ArtifactUI.GetArtifactInfo())
	if itemId == nil or itemId == 0 then
		if not artifactFrameOpen then
			HideUIPanel(ArtifactFrame)
		end
		return false, false, 0
	end

	-- if not select(1, IsUsableItem(itemId)) then
		-- if not artifactFrameOpen then
			-- HideUIPanel(ArtifactFrame)
		-- end
		-- return false, false, 0
	-- end

	local mhId = select(1, GetInventoryItemID("player", GetInventorySlotInfo("MainHandSlot")))
	local ohId = select(1, GetInventoryItemID("player", GetInventorySlotInfo("SecondaryHandSlot")))
	local correctArtifactOpen = (mhId ~= nil and mhId == itemId) or (ohId ~= nil and ohId == itemId)

	if not correctArtifactOpen then
		print("|cFFFF0000Warning, attempting to generate Simulationcraft artifact output for the wrong item (expected " .. (mhId or 0) .. " or " .. (ohId or 0) .. ", got " .. itemId .. ")")
		HideUIPanel(ArtifactFrame)
		SocketInventoryItem(INVSLOT_MAINHAND)
		itemId = select(1, ArtifactUI.GetArtifactInfo())
	end

	return artifactFrameOpen, correctArtifactOpen, itemId
end

function PersoLib:CloseArtifactFrame(wasOpen, correctOpen)
	local ArtifactFrame = _G.ArtifactFrame

	if ArtifactFrame and (not wasOpen or not correctOpen) then
		HideUIPanel(ArtifactFrame)
	end
end

-- simc, generates artifact string
function PersoLib:GetArtifactString()
	local artifactFrameOpen, correctArtifactOpen, itemId = self:OpenArtifact()

	if not itemId then
		self:CloseArtifactFrame(artifactFrameOpen, correctArtifactOpen)
		return nil
	end

	local artifactId = ExtraData.ArtifactTable[itemId]
	if artifactId == nil then
		self:CloseArtifactFrame(artifactFrameOpen, correctArtifactOpen)
		return nil
	end

	-- Note, relics are handled by the item string
	local str = artifactId .. ':0:0:0:0'

	local baseRanks = {}
	local crucibleRanks = {}

	local powers = ArtifactUI.GetPowers()
	for i = 1, #powers do
		local powerId, powerRank = PersoLib:GetPowerData(powers[i])

		if powerRank > 0 then
			baseRanks[#baseRanks + 1] = powerId
			baseRanks[#baseRanks + 1] = powerRank
		end
	end

	if #baseRanks > 0 then
		str = str .. ':' .. table.concat(baseRanks, ':')
	end

	self:CloseArtifactFrame(artifactFrameOpen, correctArtifactOpen)

	return str
end

-- simc, generates crucible string
function PersoLib:GetCrucibleString(RelicSlot)
	local artifactFrameOpen, correctArtifactOpen, itemId = self:OpenArtifact()

	if not itemId then
		self:CloseArtifactFrame(artifactFrameOpen, correctArtifactOpen)
		return nil
	end

	local artifactId = ExtraData.ArtifactTable[itemId]
	if artifactId == nil then
		self:CloseArtifactFrame(artifactFrameOpen, correctArtifactOpen)
		return nil
	end

  
	local crucibleData = {}
	for ridx = 1, ArtifactUI.GetNumRelicSlots() do
		crucibleData[ridx] = PersoLib:GetCrucibleStringForSlot(ridx)
	end

	local crucibleStrings = {}
	for ridx = 1, #crucibleData do
		crucibleStrings[ridx] = table.concat(crucibleData[ridx], ':')
	end

	self:CloseArtifactFrame(artifactFrameOpen, correctArtifactOpen)

	return table.concat(crucibleStrings, '/')
end

-- simc generates crucible string for a specific slot
function PersoLib:GetCrucibleStringForSlot(RelicSlot,ForceOpenArtifact)
	if ForceOpenArtifact then
		local artifactFrameOpen, correctArtifactOpen, itemId = self:OpenArtifact()

		if not itemId then
			self:CloseArtifactFrame(artifactFrameOpen, correctArtifactOpen)
			return nil
		end

		local artifactId = ExtraData.ArtifactTable[itemId]
		if artifactId == nil then
			self:CloseArtifactFrame(artifactFrameOpen, correctArtifactOpen)
			return nil
		end
	end

	local link = select(4, ArtifactUI.GetRelicInfo(RelicSlot))
	if link ~= nil then
		local relicSplit     = PersoLib:LinkSplit(link)
		local baseLink       = select(2, GetItemInfo(relicSplit[1]))
		local basePowers     = { ArtifactUI.GetPowersAffectedByRelicItemLink(baseLink) }
		local relicPowers    = { ArtifactUI.GetPowersAffectedByRelic(RelicSlot) }
		local cruciblePowers = {}

		for rpidx = 1, #relicPowers do
			local found = false
			for bpidx = 1, #basePowers do
				if relicPowers[rpidx] == basePowers[bpidx] then
					found = true
					break
				end
			end

			if not found then
				cruciblePowers[#cruciblePowers + 1] = relicPowers[rpidx]
			end
		end
		
		if ForceOpenArtifact then
			self:CloseArtifactFrame(artifactFrameOpen, correctArtifactOpen)
		end

		if #cruciblePowers == 0 then
			return { 0 }
		else
			return cruciblePowers
		end
	else
		if ForceOpenArtifact then
			self:CloseArtifactFrame(artifactFrameOpen, correctArtifactOpen)
		end
		return { 0 }
	end
end

function PersoLib:ExtractCurrentCrucibleTree()
	local currentTree={}
	if ArtifactRelicForgeUI.IsAtForge() then
		if _G.ArtifactRelicForgeFrame.relicSlot and _G.ArtifactRelicForgeFrame.relicSlot <=3 then--relic slot
			local treeData={}
			treeData = ArtifactRelicForgeUI.GetSocketedRelicTalents(_G.ArtifactRelicForgeFrame.relicSlot)
			for k,v in pairs(treeData) do
				if not currentTree[treeData[k].tier] then currentTree[treeData[k].tier]={} end
				table.insert(currentTree[treeData[k].tier],treeData[k].powerID)
			end
		else -- preview slot
			local treeData={}
			treeData = ArtifactRelicForgeUI.GetPreviewRelicTalents()
			for k,v in pairs(treeData) do
				if not currentTree[treeData[k].tier] then currentTree[treeData[k].tier]={} end
				table.insert(currentTree[treeData[k].tier],treeData[k].powerID)
			end
		end
	end
	return currentTree
end

-- get item id from link
function PersoLib:GetIDFromLink(itemLink)
	local itemString = string.match(itemLink, "item:([%-?%d:]+)")
	local itemSplit = {}

	-- Split data into a table
	for v in string.gmatch(itemString, "(%d*:?)") do
		if v == ":" then
		  itemSplit[#itemSplit + 1] = 0
		else
		  itemSplit[#itemSplit + 1] = string.gsub(v, ':', '')
		end
	end

	return tonumber(itemSplit[1])
end

-- get item id from link
function PersoLib:GetILVLFromLink(itemLink)
	local ilvl
	_,_,_,ilvl = GetItemInfo(itemLink)
	return ilvl
end

function PersoLib:debugPrint(str,affichedebug)
	if affichedebug then
		print(str)
	end
end

function PersoLib:MergeTables(tableDefault,tableVars,tablereception)
	for k,v in pairs(tableDefault) do
		if tableVars[k] == nil then
			tablereception[k]=v
		else
			tablereception[k]=tableVars[k]
		end
	end
	return tablereception
end

function PersoLib:GetRealIlvl(itemLink)
	local itemString = string.match(itemLink, "item:([%-?%d:]+)")
	local itemSplit = {}
	
	-- Split data into a table
	for _, v in ipairs({strsplit(":", itemString)}) do
		if v == "" then
		  itemSplit[#itemSplit + 1] = 0
		else
		  itemSplit[#itemSplit + 1] = tonumber(v)
		end
	end
	_,_,itemRarity,ilvl = GetItemInfo(itemLink)
	if itemRarity==7 then --heirloom
		ilvl = 815
	elseif tonumber(itemSplit[11])==512 and tonumber(itemSplit[12])==22 and tonumber(itemSplit[15])==110 then --timewalking
		ilvl = 850
		--todo : 895 if warforged
	end
	
	return ilvl
end

function PersoLib:LinkSplit(link)
  local itemString = string.match(link, "item:([%-?%d:]+)")
  local itemSplit = {}

  -- Split data into a table
  for _, v in ipairs({strsplit(":", itemString)}) do
    if v == "" then
      itemSplit[#itemSplit + 1] = 0
    else
      itemSplit[#itemSplit + 1] = tonumber(v)
    end
  end

  return itemSplit
end

function PersoLib:GetNameFromTraitID(traitID,artifactID)
	local name=""
	if ExtraData.NetherlightData[1][traitID] then --crucible trait
		name = ExtraData.NetherlightData[1][traitID]
	elseif  ExtraData.NetherlightData[2][traitID] then--artifact trait
		name = ExtraData.NetherlightData[2][traitID]
	elseif ExtraData.NetherlightData[3][artifactID][traitID] then
		name = ExtraData.NetherlightData[3][artifactID][traitID]
	else
		name = ""
	end
	
	return name
end

function PersoLib:DumpTable(tbl, indent)
  if not indent then indent = 0 end
  for k, v in pairs(tbl) do
    formatting = string.rep("  ", indent) .. k .. ": "
    if type(v) == "table" then
      print(formatting)
      PersoLib:DumpTable(v, indent+1)
    elseif type(v) == 'boolean' then
      print(formatting .. tostring(v))      
    else
      print(formatting .. v)
    end
  end
end

function PersoLib:Dump(tbl,tblname,inspect)
	if tblname then
		print("Dump:"..tblname)
	end
	DevTools_Dump(tbl);
	if inspect then
		UIParentLoadAddOn("Blizzard_DebugTools")
		DisplayTableInspectorWindow(tbl)
	end
end


extras.lua
local _, SimPermut = ...

-- UI
SimPermut.ExtraData={
	-- SimcData
	RegionString = {
	  [1] = 'us',
	  [2] = 'kr',
	  [3] = 'eu',
	  [4] = 'tw',
	  [5] = 'cn'
	},
	SpecNames = {
	-- Death Knight
	  [250] = 'Blood',
	  [251] = 'Frost',
	  [252] = 'Unholy',
	-- Demon Hunter
	  [577] = 'Havoc',
	  [581] = 'Vengeance',
	-- Druid 
	  [102] = 'Balance',
	  [103] = 'Feral',
	  [104] = 'Guardian',
	  [105] = 'Restoration',
	-- Hunter 
	  [253] = 'Beast Mastery',
	  [254] = 'Marksmanship',
	  [255] = 'Survival',
	-- Mage 
	  [62] = 'Arcane',
	  [63] = 'Fire',
	  [64] = 'Frost',
	-- Monk 
	  [268] = 'Brewmaster',
	  [269] = 'Windwalker',
	  [270] = 'Mistweaver',
	-- Paladin 
	  [65] = 'Holy',
	  [66] = 'Protection',
	  [70] = 'Retribution',
	-- Priest 
	  [256] = 'Discipline',
	  [257] = 'Holy',
	  [258] = 'Shadow',
	-- Rogue 
	  [259] = 'Assassination',
	  [260] = 'Outlaw',
	  [261] = 'Subtlety',
	-- Shaman 
	  [262] = 'Elemental',
	  [263] = 'Enhancement',
	  [264] = 'Restoration',
	-- Warlock 
	  [265] = 'Affliction',
	  [266] = 'Demonology',
	  [267] = 'Destruction',
	-- Warrior 
	  [71] = 'Arms',
	  [72] = 'Fury',
	  [73] = 'Protection'
	},
	RoleTable={
		-- Death Knight
		[250] = 'tank',
		[251] = 'attack',
		[252] = 'attack',
		-- Demon Hunter
		[577] = 'attack',
		[581] = 'tank',
		-- Druid
		[102] = 'spell',
		[103] = 'attack',
		[104] = 'tank',
		[105] = 'heal',
		-- Hunter
		[253] = 'attack',
		[254] = 'attack',
		[255] = 'attack',
		-- Mage
		[62] = 'spell',
		[63] = 'spell',
		[64] = 'spell',
		-- Monk
		[268] = 'tank',
		[269] = 'attack',
		[270] = 'hybrid',
		-- Paladin
		[65] = 'heal',
		[66] = 'tank',
		[70] = 'attack',
		-- Priest
		[256] = 'spell',
		[257] = 'heal',
		[258] = 'spell',
		-- Rogue
		[259] = 'attack',
		[260] = 'attack',
		[261] = 'attack',
		-- Shaman
		[262] = 'spell',
		[263] = 'attack',
		[264] = 'heal',
		-- Warlock
		[265] = 'spell',
		[266] = 'spell',
		[267] = 'spell',
		-- Warrior
		[71] = 'attack',
		[72] = 'attack',
		[73] = 'attack'
	},
	-- Permutation data
	GetItemInfoName = {
		["INVTYPE_HEAD"]=1,
		["INVTYPE_NECK"]=2,
		["INVTYPE_SHOULDER"]=3,
		["INVTYPE_CLOAK"]=4,
		["INVTYPE_CHEST"]=5,
		["INVTYPE_ROBE"]=5,
		["INVTYPE_WRIST"]=6,
		["INVTYPE_HAND"]=7,
		["INVTYPE_WAIST"]=8,
		["INVTYPE_LEGS"]=9,
		["INVTYPE_FEET"]=10,
		["INVTYPE_FINGER"]=11,
		["INVTYPE_TRINKET"]=12,
		["INVSLOT_MAINHAND"]=13,
		["INVSLOT_OFFHAND"]=14
	},
	SlotNames = {
		[1]="HeadSlot", 
		[2]="NeckSlot", 
		[3]="ShoulderSlot", 
		[4]="BackSlot", 
		[5]="ChestSlot", 
		[6]="ShirtSlot", 
		[7]="TabardSlot", 
		[8]="WristSlot", 
		[9]="HandsSlot", 
		[10]="WaistSlot", 
		[11]="LegsSlot", 
		[12]="FeetSlot", 
		[13]="Finger0Slot", 
		[14]="Finger1Slot", 
		[15]="Trinket0Slot", 
		[16]="Trinket1Slot", 
		[17]="MainHandSlot", 
		[18]="SecondaryHandSlot", 
		[19]="AmmoSlot" 
	},
	SimcSlotNames = {
		[1]="head",
		[2]="neck",
		[3]="shoulder",
		[4]="back",
		[5]="chest",
		[6]="shirt",
		[7]="tabard",
		[8]="wrist",
		[9]="hands",
		[10]="waist",
		[11]="legs",
		[12]="feet",
		[13]="finger1",
		[14]="finger2",
		[15]="trinket1",
		[16]="trinket2",
		[17]="main_hand",
		[18]="off_hand",
		[19]="ammo"
	},
	ListNames = {
		[1]="head",
		[2]="neck",
		[3]="shoulder",
		[4]="back",
		[5]="chest",
		[6]="wrist",
		[7]="hands",
		[8]="waist",
		[9]="legs",
		[10]="feet",
		[11]="finger",
		[12]="trinket",
		[13]="main hand",
		[14]="off hand"
	},
	PermutSimcNames = {
		[1]="head",
		[2]="neck",
		[3]="shoulder",
		[4]="back",
		[5]="chest",
		[6]="wrist",
		[7]="hands",
		[8]="waist",
		[9]="legs",
		[10]="feet",
		[11]="finger1",
		[12]="finger2",
		[13]="trinket1",
		[14]="trinket2",
		[15]="main_hand",
		[16]="off_hand"
	},
	PermutSlotNames = {
		[1]="HeadSlot", 
		[2]="NeckSlot", 
		[3]="ShoulderSlot", 
		[4]="BackSlot", 
		[5]="ChestSlot", 
		[6]="WristSlot", 
		[7]="HandsSlot", 
		[8]="WaistSlot", 
		[9]="LegsSlot", 
		[10]="FeetSlot", 
		[11]="Finger0Slot", 
		[12]="Finger1Slot", 
		[13]="Trinket0Slot", 
		[14]="Trinket1Slot", 
		[15]="MainHandSlot", 
		[16]="SecondaryHandSlot"
	},

	-- UI
	FrameMenu={
		[1] = "Gear permutation",
		[2] = "Talent permutation",
		[3] = "Add Items",
		[4] = "Options"
	},
	ReportTypeGear = {
		[1] = 'Item names',
		[2] = 'Copy number'
	},
	ReportTypeTalents = {
		[1] = 'Talents taken',
		[2] = 'Copy number'
	},
	enchantWeapon = {
		[0] 	= 'Untouched',
		[5946] 	= 'Coastal Surge',
		[5950] 	= 'Gale-Force Striking',
		[5949] 	= 'Torrent of Elements',
		[5948] 	= 'Siphoning',
		[5965] 	= 'Deadly Navigation',
		[5964] 	= 'Masterful Navigation',
		[5963] 	= 'Quick Navigation',
		[5966] 	= 'Stalwart Navigation',
		[5962] 	= 'Versatile Navigation',
		[999999]= 'Scrap ench.'
	},
	enchantRing = {
		[0] 	= 'Untouched',
		[5942] 	= '+37 Crit',
		[5943] 	= '+37 Haste',
		[5944] 	= '+37 Mast',
		[5945] 	= '+37 Vers',
		[999999]= 'Scrap ench.'
	},
	gemList = {
		[0] 	 = 'Untouched',
		[153710] = '+30 Crit',
		[153711] = '+30 Haste',
		[153713] = '+30 Mast',
		[153712] = '+30 Vers',
		[154126] = '+40 Crit',
		[154127] = '+40 Haste',
		[154129] = '+40 Mast',
		[154128] = '+40 Vers',
		[999999] = 'Scrap gem'
	},
	gemListEpic = {
		[0] 	   = 'No Epic gem',
		[130246] = '+40 Str',
		[130247] = '+40 Agi',
		[130248] = '+40 Int',
	}
}


